smarts/core/provider.py

```diff
@@ -76,3 +88,28 @@ def reset(self):
 
     def teardown(self):
         raise NotImplementedError
+
+    def recover(
+        self, scenario, elapsed_sim_time: float, error: Optional[Exception] = None
+    ) -> bool:
```

Comment:
I'm wondering if this shouldn't also return a new `ProviderState`, just in case recovery results in something different than the original attempt.